src/Namecheap/Provider.php

```diff
@@ -0,0 +1,502 @@
+        }
+    }
+
+    /**
+     * @param  RegisterDomainParams  $params
+     *
+     * @return void
+     */
+    private function checkRegisterParams(RegisterDomainParams $params): void
+    {
+        if (!Arr::has($params, 'registrant.register')) {
+            throw $this->errorResult('Registrant contact data is required!');
+        }
+
+        if (!Arr::has($params, 'tech.register')) {
+            throw $this->errorResult('Tech contact data is required!');
+        }
+
+        if (!Arr::has($params, 'admin.register')) {
+            throw $this->errorResult('Admin contact data is required!');
+        }
+
+        if (!Arr::has($params, 'billing.register')) {
+            throw $this->errorResult('Billing contact data is required!');
+        }
+    }
+
+
+    /**
+     * @param ArrayAccess $params
+     * @param string $prefix
+     * @return string
+     */
+    private function prepareNameservers(ArrayAccess $params, string $prefix): string
+    {
+        $nameServers = "";
+
+        $custom = 0;
+        $default = 0;
+
+        for ($i = 1; $i <= self::MAX_CUSTOM_NAMESERVERS; $i++) {
+            if (Arr::has($params, $prefix.$i)) {
+                $host = Arr::get($params, $prefix.$i)->host;
+                if (!in_array($host, self::DEFAULT_NAMESERVERS)) {
+                    $nameServers .= $host.",";
+                    $custom++;
+                } else {
+                    $default++;
+                }
+            }
+        }
+
+        if ($custom != 0 && $default != 0) {
+            throw $this->errorResult(
+                "It's not possible to mix Namecheap's default nameservers with other ones",
+                $params
+            );
+        }
+
+        if ($custom + $default < self::MIN_CUSTOM_NAMESERVERS) {
+            throw $this->errorResult('Minimum two nameservers are required!', $params);
+        }
+
+        return $nameServers;
+    }
+
+    /**
+     * @param  TransferParams  $params
+     *
+     * @return DomainResult
+     */
+    public function transfer(TransferParams $params): DomainResult
+    {
+        $sld = Utils::normalizeSld($params->sld);
+        $tld = Utils::normalizeTld($params->tld);
+
+        if (!in_array($tld, NamecheapApi::TRANSFER_TLD)) {
+            throw $this->errorResult(sprintf("Transfer is not available for TLD %s", $tld), $params);
+        }
+
+        $domainName = Utils::getDomain($sld, $tld);
+
+        $eppCode = $params->epp_code ?: '0000';
+
+        try {
+            return $this->_getInfo($domainName, 'Domain active in registrar account');
+        } catch (Throwable $e) {
+            // domain not active - continue below
+        }
+
+        try {
+            $prevOrder = $this->api()->getDomainTransferOrders($domainName);
+
+            if (is_null($prevOrder)) {
+                $transferId = $this->api()->initiateTransfer($domainName, $eppCode);
+
+                return DomainResult::create([
```

Comment:
The intended flow for the `transfer()` function is that it will be retried periodically until a success result is returned. In the next few weeks i'll refactor this in all providers to be split into 2 separate functions for initiate and check. So, for now, please throw an error result after initiating the transfer since transfers typically take several days before the domain is active in the registrar account.